Source/MLX/MLXArray.swift

```diff
@@ -246,6 +241,109 @@ public final class MLXArray {
         asType(T.dtype, stream: stream)
     }
 
+    /// Return the dimension where the storage is contiguous.
+    ///
+    /// If this returns 0 then the whole storage is contiguous.  If it returns ndmin + 1 then none of it is contiguous.
+    func contiguousToDimension() -> Int {
+        let shape = self.shape
+        let strides = self.strides
+
+        var expectedStride = 1
+
+        for (dimension, (shape, stride)) in zip(shape, strides).enumerated().reversed() {
+            // as long as the actual strides match the expected (contiguous) strides
+            // the backing is contiguous in these dimensions
+            if stride != expectedStride {
+                return dimension + 1
+            }
+            expectedStride *= shape
+        }
+
+        return 0
+    }
+
+    /// Return the physical size of the backing (assuming it is evaluated) in elements
+    var physicalSize: Int {
```

Comment:
Does this (and its usage) consider the case that a data buffer can have holes in it? For example if you do:

```
a = mx.zeros((4, 4))[:, ::2]
```

Then the strides of `a` will be `[4, 2]`. The physical size of the underlying buffer is still 16 but it has holes punched in it for every other element.

Reply:
It should, but let me add a test for that.

I wonder if this copy to contiguous should be part of mlx::core::array?  This is a lot of detailed knowledge of the underlying memory.

Reply:
Looks good:

```swift
    func testAsArrayNonContiguous4() {
        // buffer with holes (last dimension has stride of 2 and
        // thus larger storage than it physically needs)
        let a = MLXArray(0 ..< 16, [4, 4])
        let s = a[0..., .stride(by: 2)]

        let expected: [Int32] = [0, 2, 4, 6, 8, 10, 12, 14]
        assertEqual(s, MLXArray(expected, [4, 2]))
        
        XCTAssertEqual(s.strides, [4, 2])

        let s_arr = s.asArray(Int32.self)
        XCTAssertEqual(s_arr, expected)
    }
```

Reply:
Thanks for the test.

> I wonder if this copy to contiguous should be part of mlx::core::array? This is a lot of detailed knowledge of the underlying memory.

Yea have a weak story right now with getting contiguous data.  There's no great solution. It doesn't really make sense as an op (which goes in the graph) since it's an implementation detail. But it's also something that we need every once in a while. I'll give this some thought but indeed having a utility in core to ensure an already evaluated array is contiguous may be a better call.